File: m/rls/rlb/lnmod.c
/* file -  lnmod - logical name routines */
#include <stdlib.h>
#include "m:\rid\rider.h"
#include "m:\rid\mxdef.h"
#include "m:\rid\dfdef.h"
#include "m:\rid\evdef.h"
#include "m:\rid\stdef.h"
#include "m:\rid\lndef.h"
dfTctx *lnPctx = NULL;
int lnVchg = 0;
int lnVerr = 0;
dfTctx *ln_ctx()
{ if( lnVerr > lnLIM)return ( NULL );
  if ( !lnPctx) {
    lnPctx = df_ctx (_lnFIL, dfEPH_); }
  if ( lnPctx == 0) {++lnVerr ;}
  return ( lnPctx);
} 
int ln_ini()
{ return ( ln_ctx () != 0);
} 
/* code -  ln_loo - simple lookup */
char *ln_loo(
char *log )
{ dfTctx *ctx ;
  dfTdef *def ;
  if( (ctx = ln_ctx ()) == NULL)return ( NULL );
  def = ctx->Proo;
  while ( def) {
    if( st_sam (log, def->Pnam))break;
    def = def->Psuc;
  } 
  if( !def)return ( NULL );
  return ( def->Pbod);
} 
/* code -  ln_trn - full translation */
int ln_trn(
char *log ,
char *res ,
int mod )
{ char nam [mxIDT];
  char *lft ;
  char *fnd = NULL;
  int loo = 16;
  st_cln (log, nam, mxIDT-1);
  st_cln (nam, res, mxIDT-1);
  st_low (nam);
  lft = nam;
  if ( st_sam (nam, _lnLOG)) {
    if( (ev_get (_lnENV, res, mxIDT)) != 0)return ( 1 );
    st_cop (_lnDEF, res);
    return 1; }
  while ( loo--) {
    if( (lft = ln_loo (lft)) == NULL)break;
    fnd = lft;
  } 
  if( fnd == 0)return 0;
  st_cop (fnd, res);
  return 1;
} 
/* code -  ln_rev - reverse translation */
int ln_rev(
char *eqv ,
char *res ,
int nth )
{ dfTctx *ctx ;
  dfTdef *def ;
  char nam [mxIDT];
  char *fnd = NULL;
  if( (ctx = ln_ctx ()) == 0)return 0;
  st_cln (eqv, nam, mxIDT-1);
  st_low (nam);
  def = ctx->Proo;
  while ( def) {
    if ( st_sam (eqv, def->Pbod)) {
      fnd = def->Pnam;
      if( !--nth)break; }
    def = def->Psuc;
  } 
  if( nth)return 0;
  st_cop (fnd, res);
  return 1;
} 
/* code -  ln_exi - exit define/undefine sequence */
void ln_exi()
{ if ( lnVchg) {ln_upd () ;}
  if ( lnPctx) {df_dlc (lnPctx) ;}
  lnPctx = NULL;
} 
/* code -  ln_upd - update logical name table */
int ln_upd()
{ if( ! lnPctx)return;
  if( (df_wri (lnPctx)) == 0){ ++lnVerr ; return 0;}
  lnVchg = 0;
  return 1;
} 
/* code -  ln_def - define logical name */
int ln_def(
char *log ,
char *equ ,
int mod )
{ char lft [mxIDT];
  char rgt [mxIDT];
  dfTctx *ctx = ln_ctx ();
  if( ctx == 0)return 0;
  st_cln (log, lft, mxIDT-1);
  st_cln (equ, rgt, mxIDT-1);
  st_low (lft);
  st_low (rgt);
  df_ins (ctx, log, equ);
  return ( ln_upd ());
} 
/* code -  ln_und - undefine logical name */
int ln_und(
char *log ,
int mod )
{ char lft [mxIDT];
  dfTctx *ctx = ln_ctx ();
  dfTdef *def ;
  char *nam ;
  int wri = 0;
  if( !ctx)return 0;
  st_cln (log, lft, mxIDT-1);
  st_low (lft);
  def = df_loo (ctx, lft);
  if( !def)return 0;
  df_del (ctx, def);
  return ( ln_upd ());
} 

File: m/rls/lnmod.r
file	lnmod - logical name routines
include <stdlib.h>
include	rid:rider
include	rid:drdef
include	rid:evdef
include	rid:mxdef
include	rid:stdef
include	rid:rgdef
include	rid:lndef

  proc	ln_exi
  is	nothing
  end

code	ln_trn - full translation

  func	ln_trn
	log : * char			; name to translate
	res : * char			; result buffer (up to mxSPC)
	mod : int			; 0=full, 1=single step
	()  : int			; fine/fail
  is	nam : [mxSPC] char		; private copy
	pen : [mxSPC] char		;
	prv : [mxSPC] char		; translated name
	fnd : int = 0			;
	loo : int = 16			;
					;
	st_fit (log, nam, mxSPC)	; get a copy
	st_low (nam)			; lower it
	st_cop (nam, res)		; default result is logical name
	prv[0] = 0			;
					;
      while loo--			; avoid infinity
	if nam[1] eq ':'		; device name (e.g. "c:")?
	.. fine st_cop (nam, res)	; yep, that's fixed
					;
	rg_get (nam, nam, mxSPC)	; translation step
	quit if fail			; no more translations
	st_cop (prv, pen)		; lazy
	st_cop (nam, prv)		;
	++fnd				; found something
	quit if mod eq 1
      end				;
	fail if !fnd			; found nothing
	(*pen && (mod & lnPEN_)) ? pen ?? prv
	fine st_cop (that, res)		; return result
   end
code	ln_def - define logical name

  func	ln_def
	log : * char
	equ : * char
	mod : int			; currently unused
	()  : int
  is	rg_set (log, equ)		; change the name
	if st_sam (log, "default")	; for "default"
	   ev_set ("Default", log)	; notify other processes
	.. dr_set (log, 0)		; change process default
	fine
  end

code	ln_und - undefine logical name

  func	ln_und
	log : * char
	mod : int			; currently unused
	()  : int
  is	rg_und (log)
  end

code	ln_nth enumerate names

  func	ln_nth
	nth : int
	log : * char
	equ : * char
	len : int
  is	fail if !rg_nth (nth, log, len)
	fine if rg_get (log, equ, len)
	fine st_cop ("?", equ)
  end

code	ln_rev - reverse translation

;	This routine translates a name in reverse.
;	The nth occurrence of the translation may be selected

  func	ln_rev
	equ : * char			; name to translate
	res : * char			; result buffer
	nth : int			; nth translation
	()  : int			; fine/fail
  is	nam : [mxIDT] char		; equ in lower case
	log : [mxIDT] char		;
	trn : [mxIDT] char		;
	idx : int = 0			; enumeration index
	st_fit (equ, nam, mxIDT)	; get a copy
	st_low (nam)			;
	repeat				;
	   ln_nth (idx++,log,trn,mxIDT)	; get next
	   pass fail			; not found
	   if st_sam (nam,trn) && !--nth; need nth instance
	   .. fine st_cop (log, res)	; got it
	forever				;
   end
end file
code	ln_dev - check device name change

;	This routine monitors changes to device names.
;	It's used by V11 to catch changes to the definitions
;	of LD0..LD7 etc.
;
;	Also called by kus:roll.r

  func	ln_dev
	nam : * char
	sig : int
  is	len : int = st_len (nam)
	if len gt 3
	|| len lt 2
	|| !ct_alp (*nam++)
	|| !ct_alp (*nam++)
	|| (*nam && !ct_dig (*nam))
	.. fail
	ev_sig (evDEV) if sig
	fine
  end
code	ln_upd - update logical name table

  func	ln_upd
	()  : int
  is	exit if !lnPctx			; nothing open
	ev_sig (evLOG)			; signal change
	df_wri (lnPctx)			; update disk copy
	pass fine			; honky dory
	fail ++lnVerr			; count it
  end
;	earlier .def file implementation
file	lnmod - logical name routines
include	rid:rider
include	rid:dfdef
include	rid:mxdef
include	rid:stdef
include	rid:lndef

	ln_ctx	: (void) *dfTctx
	lnPctx : * dfTctx extern

code	ln_exi - exit define/undefine sequence

  proc	ln_exi
  is	ln_upd () if lnVchg		; update if required
	df_dlc (lnPctx) if lnPctx	; deallocate the lot
	lnPctx = <>			;
  end

code	ln_upd - update logical name table

  func	ln_upd
	()  : int
  is	exit if not lnPctx		; nothing open
	df_wri (lnPctx)			; update disk copy
	fail ++lnVerr if fail		; count it
	lnVchg = 0			; disk copy is updated
	fine				;
  end

code	ln_def - define logical name

  func	ln_def
	log : * char
	equ : * char
	mod : int			; currently unused
	()  : int
  is	lft : [mxIDT] char
	rgt : [mxIDT] char
	ctx : * dfTctx = ln_ctx ()
	fail if ctx eq
	st_cln (log, lft, mxIDT-1)
	st_cln (equ, rgt, mxIDT-1)
	st_low (lft)
	st_low (rgt)
	df_ins (ctx, log, equ)
	reply ln_upd ()
  end

code	ln_und - undefine logical name

  func	ln_und
	log : * char
	mod : int			; currently unused
	()  : int
  is	lft : [mxIDT] char
	ctx : * dfTctx = ln_ctx ()
	def : * dfTdef
	nam : * char
	wri : int = 0
	fail if !ctx
	st_cln (log, lft, mxIDT-1)
	st_low (lft)
	def = df_loo (ctx, lft)
	fail if !def
	df_del (ctx, def)
	reply ln_upd ()
  end
